sim4rec/response/nn_response.py

```diff
@@ -0,0 +1,202 @@
+        StructField("response", IntegerType(), True),
+        StructField("__iter", IntegerType(), True),
+    ]
+)
+SIM_LOG_COLS = [field.name for field in SIM_LOG_SCHEMA.fields]
+
+
+class NNResponseTransformer(ActionModelTransformer):
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.hist_data = None
+        for param, value in kwargs.items():
+            setattr(self, param, value)
+
+    @classmethod
+    def load(cls, checkpoint_dir):
+        with open(os.path.join(checkpoint_dir, "_params.pkl"), "rb") as f:
+            params_dict = pickle.load(f)
+        params_dict["backbone_response_model"] = ResponseModel.load(checkpoint_dir)
+        with open(os.path.join(checkpoint_dir, "_item_indexer.pkl"), "rb") as f:
+            params_dict["item_indexer"] = pickle.load(f)
+        with open(os.path.join(checkpoint_dir, "_user_indexer.pkl"), "rb") as f:
+            params_dict["user_indexer"] = pickle.load(f)
+        return cls(**params_dict)
+
+    def save(self, path):
+        """Save model at given path."""
+        os.makedirs(path)
+        self.backbone_response_model.dump(path)
+        with open(os.path.join(path, "_item_indexer.pkl"), "wb") as f:
+            pickle.dump(self.item_indexer, f, pickle.HIGHEST_PROTOCOL)
+        with open(os.path.join(path, "_user_indexer.pkl"), "wb") as f:
+            pickle.dump(self.user_indexer, f, pickle.HIGHEST_PROTOCOL)
+        with open(os.path.join(path, "_params.pkl"), "wb") as f:
+            pickle.dump(
+                {
+                    "outputCol": self.outputCol,
+                    "log_dir": self.log_dir,
+                    "hist_data_dir": self.hist_data_dir,
+                },
+                f,
+                pickle.HIGHEST_PROTOCOL,
+            )
+
+    def _transform(self, new_recs):
+        """
+        Predict responses for given dataframe with recommendations.
+
+        :param dataframe: new recommendations.
+        """
+
+        def predict_udf(df):
+            # if not do this, something unstable happens to the Method Resolution Order
+            from .nn_utils.datasets import PandasRecommendationData
+
+            dataset = PandasRecommendationData(
+                log=df,
+                item_indexer=self.item_indexer,
+                user_indexer=self.user_indexer,
+            )
+
+            # replacing clicks in datset with predicted
+            dataset = self.backbone_response_model.transform(dataset=dataset)
+
+            return dataset._log[SIM_LOG_COLS]
+
+        spark = new_recs.sql_ctx.sparkSession
+
+        # read the historical data
+        hist_data = spark.read.schema(SIM_LOG_SCHEMA).parquet(self.hist_data_dir)
+        if not hist_data:
+            print("Warning: the historical data is empty")
+            hist_data = spark.createDataFrame([], schema=SIM_LOG_SCHEMA)
+        # filter users whom we don't need
+        hist_data = hist_data.join(new_recs, on="user_idx", how="semi")
```

Comment:
If you really want to leave the history of only distinct users from `new_recs` in `hist_data`.

```suggestion
        hist_data = hist_data.join(sf.broadcast(new_recs.select("user_idx").distinct()), on="user_idx", how="inner")
```